Allow MessageEncryptor to take advantage authenticated encryption modes

bdewater opened the pull request:
AEAD modes have been [supported since Ruby 2.0](https://bugs.ruby-lang.org/issues/6980) ([doc](http://ruby-doc.org/stdlib-2.3.0/libdoc/openssl/rdoc/OpenSSL/Cipher.html#class-OpenSSL::Cipher-label-Authenticated+Encryption+and+Associated+Data+-28AEAD-29)) and OpenSSL 1.0.1. It is faster:

``` ruby
require 'benchmark/ips'
require 'active_support/message_encryptor'
require 'active_support/key_generator'

Benchmark.ips do |x|
  salt  = SecureRandom.random_bytes(64)
  key   = ActiveSupport::KeyGenerator.new('password').generate_key(salt)
  data  = 'my secret data'

  x.report("default roundtrip") do
    crypt = ActiveSupport::MessageEncryptor.new(key)
    encrypted_data = crypt.encrypt_and_sign(data)
    crypt.decrypt_and_verify(encrypted_data)
  end

  x.report("aes-256-gcm roundtrip") do
    crypt = ActiveSupport::MessageEncryptor.new(key, cipher: 'aes-256-gcm')
    encrypted_data = crypt.encrypt_and_sign(data)
    crypt.decrypt_and_verify(encrypted_data)
  end

  x.compare!
end
```

```
Warming up --------------------------------------
   default roundtrip     2.148k i/100ms
aes-256-gcm roundtrip
                         4.698k i/100ms
Calculating -------------------------------------
   default roundtrip     21.947k (± 3.4%) i/s -    111.696k in   5.095390s
aes-256-gcm roundtrip
                         48.374k (± 4.2%) i/s -    244.296k in   5.058709s

Comparison:
aes-256-gcm roundtrip:    48374.4 i/s
   default roundtrip:    21946.7 i/s - 2.20x slower
```

And it produces smaller ciphertexts:

```
irb(main):090:0> ActiveSupport::MessageEncryptor.new(key).encrypt_and_sign(data).length
=> 138
irb(main):091:0> ActiveSupport::MessageEncryptor.new(key, cipher: 'aes-256-gcm').encrypt_and_sign(data).length
=> 76
```

bdewater committed:
Allow MessageEncryptor to take advantage authenticated encryption modes

AEAD modes like `aes-256-gcm` provide both confidentiality and data authenticity, eliminating the need to use MessageVerifier to check if the encrypted data has been tampered with.

rafaelfranca commented:
r? @matthewd @jeremy

jeremy commented:
Dig it! Can you ensure compatibility with MessageEncryptor that already use an AEAD cipher?

bdewater commented:
That could probably work by counting the `--` occurrences, I'll work on it somewhere later this week.

bdewater committed:
The OpenSSL bindings can tell us if a cipher supports AEAD

Using `aes-256-ccm` as a cipher on Ruby 2.3.1 results in "OpenSSL::Cipher::CipherError: authentication tag not supported by this cipher" even though it is an authenticated mode: https://wiki.openssl.org/index.php/EVP_Authenticated_Encryption_and_Decryption#Authenticated_Encryption_using_CCM_mode - the same will probably be true for `chacha20-poly1305` when OpenSSL 1.1.0 is released without a fix upstream in the bindings..

bdewater commented:
@jeremy that should not work right now. MessageVerifier will tell that the message is valid but `OpenSSL::Cipher::CipherError` will be raised during decryption because the `auth_tag` is not set.

I've also done a quick test with [aes-256-ccm](https://wiki.openssl.org/index.php/EVP_Authenticated_Encryption_and_Decryption#Authenticated_Encryption_using_CCM_mode) (the only AEAD chipher on my system with OpenSSL 1.0.2h according to `OpenSSL::Cipher.ciphers`) on my branch and it looks like that the Ruby bindings do not support this mode as authenticated. I've changed the `#aead_mode?` check to use what the bindings report about the cipher.

bdewater committed:
Raise error if the auth_tag has been shortened, see https://github.com/ruby/openssl/issues/63

jeremy commented on activesupport/CHANGELOG.md:
```diff
@@ -1,3 +1,12 @@
+*   Allow MessageEncryptor to take advantage authenticated encryption modes.
```
take advantage *of

jeremy commented on activesupport/lib/active_support/message_encryptor.rb:
```diff
@@ -28,6 +29,16 @@ def self.dump(value)
       end
     end
 
+    module AeadVerifier #:nodoc:
```
This is used in tandem with AEAD ciphers, but it isn't an AEAD verifier—it's a no-op verifier.

bdewater committed:
Fix some wording

bdewater commented:
Thanks for the feedback @jeremy, PR updated! However #23297 is also on the 5.1.0 milestone, in which direction does the team want to go with MessageEncryptor?

bdewater committed:
Remove last byte instead of hardcoding length

jeremy commented:
Squashed in d4ea18a8cb84601509ee4c6dc691b212af8c2c36

jeremy commented:
Thanks @bdewater! #23297 is a much larger effort to switch to a new message envelope (and more). Your work is independent and good for release even if the JWE work slips to 5.2.

vipulnsward commented:
We should be adding a note that, switching modes busts previous cookies/auth that were set with current environment.

vipulnsward commented:
Hmm, nevermind, I don't see how this will affect existing auth
